File: string-paradigm-expression-no-wrapper/src/main/java/org/jesperancinha/string/paradigms/nowrapper/expression/SpellingCheckBase.java
package org.jesperancinha.string.paradigms.nowrapper.expression;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class SpellingCheckBase {

    protected final List<Object> spellingDoubleCommaRegions = new ArrayList<Object>();
    protected final List<Object> spellingWhiteSpacesRegions = new ArrayList<Object>();
    protected final List<Object> spellingBeginCapitalRegions = new ArrayList<Object>();
    
    private static final Pattern DOUBLE_COMMAS = Pattern.compile(".*,\\s*,.*");
    private static final Pattern WHITE_SPACES = Pattern.compile(".*\\s\\s+.*");

    protected void checkSpelling(Object component) {
    	if (DOUBLE_COMMAS.matcher(component.toString()).matches()) {
    	    spellingDoubleCommaRegions.add(component);
    	}
    	if (WHITE_SPACES.matcher(component.toString()).matches()) {
    	    spellingWhiteSpacesRegions.add(component);
    	}
    	String[] elements = component.toString().split(",");
    	for (int i = 0; i < elements.length; i++) {
    		String element = elements[i].trim();
    		if (element.length() > 0 && Character.getType(element.charAt(0)) == Character.LOWERCASE_LETTER) {
    		    spellingBeginCapitalRegions.add(component);
    		}
    	}
    }
}

File: jeorg-kotlin-apps/string-array-paradigms/string-paradigm-expression_1/src/main/kotlin/org/jesperancinha/string/paradigm1/expression/SpellingCheckBase.kt
package org.jesperancinha.string.paradigm1.expression

import java.util.regex.Pattern

open class SpellingCheckBase<T> {
    val spellingDoubleCommaRegions: MutableList<T> = ArrayList()
    val spellingWhiteSpacesRegions: MutableList<T> = ArrayList()
    val spellingBeginCapitalRegions: MutableList<T> = ArrayList()
    fun checkSpelling(component: T) {
        if (DOUBLE_COMMAS.matcher(component.toString()).matches()) {
            spellingDoubleCommaRegions.add(component)
        }
        if (WHITE_SPACES.matcher(component.toString()).matches()) {
            spellingWhiteSpacesRegions.add(component)
        }
        val elements = component.toString().split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (i in elements.indices) {
            val element = elements[i].trim { it <= ' ' }
            if (element.length > 0 && Character.getType(element[0]) == Character.LOWERCASE_LETTER.toInt()) {
                spellingBeginCapitalRegions.add(component)
            }
        }
    }

    companion object {
        private val DOUBLE_COMMAS = Pattern.compile(".*,\\s*,.*")
        private val WHITE_SPACES = Pattern.compile(".*\\s\\s+.*")
    }
}